File: app/src/main/java/com/blogspot/e_kanivets/moneytracker/activity/account/EditAccountActivity.java
package com.blogspot.e_kanivets.moneytracker.activity.account;

import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.EditText;

import com.blogspot.e_kanivets.moneytracker.R;
import com.blogspot.e_kanivets.moneytracker.activity.base.BaseBackActivity;
import com.blogspot.e_kanivets.moneytracker.controller.data.AccountController;
import com.blogspot.e_kanivets.moneytracker.entity.data.Account;

import javax.inject.Inject;

import butterknife.Bind;
import butterknife.OnClick;

public class EditAccountActivity extends BaseBackActivity {

    public static final String KEY_ACCOUNT = "key_account";

    @Inject
    AccountController accountController;

    private Account account;

    @Bind(R.id.et_title)
    EditText etTitle;
    @Bind(R.id.et_goal)
    EditText etGoal;
    @Bind(R.id.view_color)
    View viewColor;

    @Override
    protected int getContentViewId() {
        return R.layout.activity_edit_account;
    }

    @Override
    protected boolean initData() {
        getAppComponent().inject(EditAccountActivity.this);
        account = getIntent().getParcelableExtra(KEY_ACCOUNT);
        return account != null && super.initData();
    }

    @Override
    protected void initViews() {
        super.initViews();

        etTitle.setText(account.getTitle());
        etGoal.setText(Double.toString(account.getGoal()));
        viewColor.setBackgroundColor(account.getColor());
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        if (account.isArchived()) {
            getMenuInflater().inflate(R.menu.menu_archived_account, menu);
        } else {
            getMenuInflater().inflate(R.menu.menu_account, menu);
        }
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_archive:
                archive();
                return true;

            case R.id.action_restore:
                restore();
                return true;

            case R.id.action_delete:
                delete();
                return true;

            default:
                return super.onOptionsItemSelected(item);
        }
    }

    @OnClick(R.id.fabDone)
    void done() {

    }

    private void archive() {
        if (account.equals(accountController.readDefaultAccount())) {
            showToast("You can't archive a default account.");
        } else {
            accountController.archive(account);
            setResult(RESULT_OK);
            finish();
        }
    }

    private void restore() {
        accountController.restore(account);
        setResult(RESULT_OK);
        finish();
    }

    private void delete() {
        accountController.delete(account);
        setResult(RESULT_OK);
        finish();
    }

}

File: app/src/main/java/com/blogspot/e_kanivets/moneytracker/activity/account/edit/EditAccountActivity.kt
package com.blogspot.e_kanivets.moneytracker.activity.account.edit

import android.app.Activity
import android.app.AlertDialog
import android.content.Context
import android.content.Intent
import android.support.v4.view.ViewPager.OnPageChangeListener
import android.view.Menu
import android.view.MenuItem
import com.blogspot.e_kanivets.moneytracker.R
import com.blogspot.e_kanivets.moneytracker.activity.account.edit.fragment.AccountOperationsFragment
import com.blogspot.e_kanivets.moneytracker.activity.account.edit.fragment.EditAccountFragment
import com.blogspot.e_kanivets.moneytracker.activity.base.BaseBackActivity
import com.blogspot.e_kanivets.moneytracker.adapter.GeneralViewPagerAdapter
import com.blogspot.e_kanivets.moneytracker.controller.data.AccountController
import com.blogspot.e_kanivets.moneytracker.entity.data.Account
import kotlinx.android.synthetic.main.activity_edit_account.fabDone
import kotlinx.android.synthetic.main.activity_edit_account.tabLayout
import kotlinx.android.synthetic.main.activity_edit_account.viewPager
import javax.inject.Inject

class EditAccountActivity : BaseBackActivity() {

    @Inject
    internal lateinit var accountController: AccountController

    private lateinit var account: Account

    override fun getContentViewId(): Int = R.layout.activity_edit_account

    override fun initData(): Boolean {
        appComponent.inject(this@EditAccountActivity)
        val accountFromParcel: Account? = intent.getParcelableExtra(KEY_ACCOUNT)

        return if (accountFromParcel == null) false
        else {
            account = accountFromParcel
            super.initData()
        }
    }

    override fun initViews() {
        super.initViews()

        tabLayout.setupWithViewPager(viewPager)

        val adapter = GeneralViewPagerAdapter(supportFragmentManager)
        adapter.addFragment(EditAccountFragment.newInstance(account), getString(R.string.information))
        adapter.addFragment(AccountOperationsFragment.newInstance(account), getString(R.string.operations))
        viewPager.adapter = adapter

        viewPager.addOnPageChangeListener(object : OnPageChangeListener {
            override fun onPageScrollStateChanged(state: Int) {}

            override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {}

            override fun onPageSelected(position: Int) {
                if (position == 0) fabDone.show() else fabDone.hide()
            }

        })
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(if (account.isArchived) R.menu.menu_archived_account else R.menu.menu_account, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.action_archive -> archive()
            R.id.action_restore -> restore()
            R.id.action_delete -> delete()
            else -> super.onOptionsItemSelected(item)
        }
    }

    private fun archive(): Boolean {
        if (account == accountController.readDefaultAccount()) {
            showToast(R.string.cant_archive_default_account)
        } else {
            accountController.archive(account)
            setResult(Activity.RESULT_OK)
            finish()
        }

        return true
    }

    private fun restore(): Boolean {
        accountController.restore(account)
        setResult(Activity.RESULT_OK)
        finish()

        return true
    }

    private fun delete(): Boolean {
        val builder = AlertDialog.Builder(this)
        builder.setTitle(R.string.delete_account_title)
        builder.setMessage(R.string.delete_account_message)
        builder.setPositiveButton(android.R.string.ok) { _, _ ->
            accountController.delete(account)
            setResult(Activity.RESULT_OK)
            finish()
        }
        builder.setNegativeButton(android.R.string.cancel, null)
        builder.show()

        return true
    }

    companion object {

        private const val KEY_ACCOUNT = "key_account"

        fun newIntent(context: Context, account: Account): Intent {
            val intent = Intent(context, EditAccountActivity::class.java)
            intent.putExtra(KEY_ACCOUNT, account)
            return intent
        }
    }

}
